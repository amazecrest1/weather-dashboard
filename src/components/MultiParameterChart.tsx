import React, { useState, useRef, useEffect } from 'react';
import { LineChart, Line, XAxis, YAxis, CartesianGrid, Legend, ResponsiveContainer } from 'recharts';
import { MultiParameterChartData, ParameterConfig } from '../types/weather';
import { HOURLY_PARAMETERS } from '../constants/cities';
import { useTheme } from '../hooks/useTheme';

interface MultiParameterChartProps {
  data: MultiParameterChartData[];
  selectedParameters: string[];
  className?: string;
}

const MultiParameterChart = ({ 
  data, 
  selectedParameters, 
  className = '' 
}: MultiParameterChartProps) => {
  const { theme } = useTheme();
  const [hoveredPoint, setHoveredPoint] = useState<{ x: number; y: number; data: any } | null>(null);
  const chartRef = useRef<HTMLDivElement>(null);
  
  // Debug: Log the current theme
  console.log('Current theme:', theme);
  console.log('Document classes:', document.documentElement.classList.toString());
  
  const getParameterConfig = (key: string): ParameterConfig | undefined => {
    return HOURLY_PARAMETERS.find(p => p.key === key);
  };

  const hasDualAxis = selectedParameters.length === 2;
  const param1 = getParameterConfig(selectedParameters[0]);
  const param2 = getParameterConfig(selectedParameters[1]);

  // Determine Y-axis assignment
  const useRightAxis = hasDualAxis && param1 && param2 && param1.yAxisId !== param2.yAxisId;

  useEffect(() => {
    const handleMouseLeave = () => {
      setHoveredPoint(null);
    };

    const chartElement = chartRef.current;
    if (chartElement) {
      chartElement.addEventListener('mouseleave', handleMouseLeave);
      return () => {
        chartElement.removeEventListener('mouseleave', handleMouseLeave);
      };
    }
  }, []);

  const handleChartMouseMove = (e: any) => {
    // Only show tooltip if we have valid data
    if (e && e.activePayload && e.activePayload.length > 0 && e.activeCoordinate) {
      setHoveredPoint({
        x: e.activeCoordinate.x,
        y: e.activeCoordinate.y,
        data: e.activePayload[0].payload
      });
    } else {
      setHoveredPoint(null);
    }
  };

  return (
    <div className={`card p-6 ${className}`}>
      <h3 className="text-lg font-semibold text-gray-900 dark:text-white mb-4">
        Hourly Weather Parameters
      </h3>
      <div 
        ref={chartRef}
        className={`rounded-lg p-4 relative ${
          theme === 'dark' 
            ? 'bg-gray-800 border border-gray-700' 
            : 'bg-white border border-gray-200'
        }`}
      >
        <ResponsiveContainer width="100%" height={400}>
        <LineChart 
          data={data}
          margin={{ top: 20, right: 30, left: 20, bottom: 20 }}
          key={`chart-${selectedParameters.join('-')}-${data.length}`}
          onMouseMove={handleChartMouseMove}
          onMouseLeave={() => setHoveredPoint(null)}
        >
          <CartesianGrid 
            strokeDasharray="3 3" 
            stroke={theme === 'dark' ? '#4B5563' : '#E5E7EB'} 
          />
          <XAxis 
            dataKey="time" 
            tick={{ 
              fontSize: 12, 
              fill: theme === 'dark' ? '#9CA3AF' : '#6B7280' 
            }}
            angle={-45}
            textAnchor="end"
            height={80}
            interval="preserveStartEnd"
            stroke={theme === 'dark' ? '#6B7280' : '#D1D5DB'}
          />
          <YAxis 
            yAxisId="left"
            orientation="left"
            tick={{ 
              fontSize: 12, 
              fill: theme === 'dark' ? '#9CA3AF' : '#6B7280' 
            }}
            stroke={theme === 'dark' ? '#6B7280' : '#D1D5DB'}
            label={{ 
              value: param1 ? `${param1.label} (${param1.unit})` : '', 
              angle: -90, 
              position: 'insideLeft',
              fill: theme === 'dark' ? '#9CA3AF' : '#6B7280'
            }}
          />
          {useRightAxis && (
            <YAxis 
              yAxisId="right"
              orientation="right"
              tick={{ 
                fontSize: 12, 
                fill: theme === 'dark' ? '#9CA3AF' : '#6B7280' 
              }}
              stroke={theme === 'dark' ? '#6B7280' : '#D1D5DB'}
              label={{ 
                value: param2 ? `${param2.label} (${param2.unit})` : '', 
                angle: 90, 
                position: 'insideRight',
                fill: theme === 'dark' ? '#9CA3AF' : '#6B7280'
              }}
            />
          )}
          <Legend 
            wrapperStyle={{
              color: theme === 'dark' ? '#F9FAFB' : '#111827'
            }}
          />
          {selectedParameters.map((paramKey, index) => {
            const config = getParameterConfig(paramKey);
            if (!config) return null;

            const yAxisId = useRightAxis && index === 1 ? 'right' : 'left';

            return (
              <Line
                key={paramKey}
                type="monotone"
                dataKey={paramKey}
                stroke={config.color}
                strokeWidth={2.5}
                name={config.label}
                yAxisId={yAxisId}
                dot={false}
                connectNulls={false}
                activeDot={false}
                isAnimationActive={false}
              />
            );
          })}
        </LineChart>
        </ResponsiveContainer>
        
        {/* Custom Tooltip */}
        {hoveredPoint && (
          <div
            style={{
              position: 'absolute',
              left: hoveredPoint.x + 10,
              top: hoveredPoint.y - 40,
              backgroundColor: theme === 'dark' ? '#1F2937' : '#FFFFFF',
              border: theme === 'dark' ? '1px solid #374151' : '1px solid #E5E7EB',
              borderRadius: '6px',
              padding: '8px 12px',
              color: theme === 'dark' ? '#F9FAFB' : '#111827',
              boxShadow: '0 4px 6px -1px rgba(0, 0, 0, 0.1)',
              fontSize: '12px',
              zIndex: 1000,
              pointerEvents: 'none',
              maxWidth: '200px'
            }}
          >
            <div style={{ margin: '0 0 4px 0', fontWeight: 'bold' }}>
              Time: {hoveredPoint.data.time}
            </div>
            {selectedParameters.map((paramKey) => {
              const config = getParameterConfig(paramKey);
              const value = hoveredPoint.data[paramKey];
              if (value === undefined || value === null) return null;
              return (
                <div key={paramKey} style={{ margin: '2px 0', color: config?.color }}>
                  {config?.label}: {value}{config?.unit}
                </div>
              );
            })}
          </div>
        )}
        
        {selectedParameters.length === 0 && (
          <div className="flex items-center justify-center h-64 text-gray-500 dark:text-gray-400">
            Please select at least one parameter to display the chart
          </div>
        )}
      </div>
    </div>
  );
};

export default MultiParameterChart; 