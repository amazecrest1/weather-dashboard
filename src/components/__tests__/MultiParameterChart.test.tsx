import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import MultiParameterChart from '../MultiParameterChart';

// Mock useTheme hook
jest.mock('../../hooks/useTheme', () => ({
  useTheme: () => ({
    theme: 'light',
    toggleTheme: jest.fn(),
    setTheme: jest.fn(),
  }),
}));

jest.mock('recharts', () => ({
  LineChart: ({ children }: { children: React.ReactNode }) => <div data-testid="line-chart">{children}</div>,
  Line: ({ name }: { name: string }) => <div data-testid={`line-${name}`}>{name}</div>,
  XAxis: () => <div data-testid="x-axis">XAxis</div>,
  YAxis: () => <div data-testid="y-axis">YAxis</div>,
  CartesianGrid: () => <div data-testid="cartesian-grid">CartesianGrid</div>,
  Tooltip: () => <div data-testid="tooltip">Tooltip</div>,
  Legend: () => <div data-testid="legend">Legend</div>,
  ResponsiveContainer: ({ children }: { children: React.ReactNode }) => (
    <div data-testid="responsive-container">{children}</div>
  ),
}));

describe('MultiParameterChart', () => {
  const mockData = [
    { time: '2024-01-01T00:00', temperature_2m: 10, wind_speed_10m: 5 },
    { time: '2024-01-01T01:00', temperature_2m: 12, wind_speed_10m: 6 },
  ];

  it('renders with one parameter', () => {
    render(<MultiParameterChart data={mockData} selectedParameters={['temperature_2m']} />);
    expect(screen.getByTestId('responsive-container')).toBeInTheDocument();
    expect(screen.getByTestId('line-chart')).toBeInTheDocument();
    expect(screen.getByTestId('x-axis')).toBeInTheDocument();
    expect(screen.getByTestId('y-axis')).toBeInTheDocument();
    expect(screen.getByTestId('legend')).toBeInTheDocument();
  });

  it('renders with two parameters', () => {
    render(<MultiParameterChart data={mockData} selectedParameters={['temperature_2m', 'wind_speed_10m']} />);
    expect(screen.getByTestId('responsive-container')).toBeInTheDocument();
    expect(screen.getByTestId('line-chart')).toBeInTheDocument();
    expect(screen.getAllByTestId('y-axis').length).toBeGreaterThanOrEqual(1);
    expect(screen.getByTestId('legend')).toBeInTheDocument();
  });

  it('renders with empty data', () => {
    render(<MultiParameterChart data={[]} selectedParameters={['temperature_2m']} />);
    expect(screen.getByTestId('responsive-container')).toBeInTheDocument();
    expect(screen.getByTestId('line-chart')).toBeInTheDocument();
  });

  it('matches snapshot with one parameter', () => {
    const { container } = render(<MultiParameterChart data={mockData} selectedParameters={['temperature_2m']} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with two parameters', () => {
    const { container } = render(<MultiParameterChart data={mockData} selectedParameters={['temperature_2m', 'wind_speed_10m']} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with empty data', () => {
    const { container } = render(<MultiParameterChart data={[]} selectedParameters={['temperature_2m']} />);
    expect(container.firstChild).toMatchSnapshot();
  });
}); 