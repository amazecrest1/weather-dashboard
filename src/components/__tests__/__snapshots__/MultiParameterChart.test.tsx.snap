// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiParameterChart matches snapshot with empty data 1`] = `
<div
  class="card p-2 sm:p-6 "
>
  <h3
    class="text-base sm:text-lg font-semibold text-gray-900 dark:text-white mb-2 sm:mb-4 flex items-center"
  >
    <svg
      class="w-5 h-5 mr-2 text-blue-500 dark:text-blue-400"
      fill="none"
      stroke="currentColor"
      viewBox="0 0 24 24"
    >
      <path
        d="M9 19v-6a2 2 0 00-2-2H5a2 2 0 00-2 2v6a2 2 0 002 2h2a2 2 0 002-2zm0 0V9a2 2 0 012-2h2a2 2 0 012 2v10m-6 0a2 2 0 002 2h2a2 2 0 002-2m0 0V5a2 2 0 012-2h2a2 2 0 012 2v14a2 2 0 01-2 2h-2a2 2 0 01-2-2z"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
      />
    </svg>
    Hourly Weather Parameters
  </h3>
  <div
    class="rounded-xl p-1 sm:p-4 relative bg-gradient-to-br from-white via-gray-50 to-white border border-gray-200 shadow-lg backdrop-blur-sm"
    style="min-height: 400px; height: 400px;"
  >
    <div
      data-testid="responsive-container"
    >
      <div
        data-testid="line-chart"
      >
        <div
          data-testid="cartesian-grid"
        >
          CartesianGrid
        </div>
        <div
          data-testid="x-axis"
        >
          XAxis
        </div>
        <div
          data-testid="y-axis"
        >
          YAxis
        </div>
        <div
          data-testid="legend"
        >
          Legend
        </div>
        <div
          data-testid="line-Temperature"
        >
          Temperature
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MultiParameterChart matches snapshot with one parameter 1`] = `
<div
  class="card p-2 sm:p-6 "
>
  <h3
    class="text-base sm:text-lg font-semibold text-gray-900 dark:text-white mb-2 sm:mb-4 flex items-center"
  >
    <svg
      class="w-5 h-5 mr-2 text-blue-500 dark:text-blue-400"
      fill="none"
      stroke="currentColor"
      viewBox="0 0 24 24"
    >
      <path
        d="M9 19v-6a2 2 0 00-2-2H5a2 2 0 00-2 2v6a2 2 0 002 2h2a2 2 0 002-2zm0 0V9a2 2 0 012-2h2a2 2 0 012 2v10m-6 0a2 2 0 002 2h2a2 2 0 002-2m0 0V5a2 2 0 012-2h2a2 2 0 012 2v14a2 2 0 01-2 2h-2a2 2 0 01-2-2z"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
      />
    </svg>
    Hourly Weather Parameters
  </h3>
  <div
    class="rounded-xl p-1 sm:p-4 relative bg-gradient-to-br from-white via-gray-50 to-white border border-gray-200 shadow-lg backdrop-blur-sm"
    style="min-height: 400px; height: 400px;"
  >
    <div
      data-testid="responsive-container"
    >
      <div
        data-testid="line-chart"
      >
        <div
          data-testid="cartesian-grid"
        >
          CartesianGrid
        </div>
        <div
          data-testid="x-axis"
        >
          XAxis
        </div>
        <div
          data-testid="y-axis"
        >
          YAxis
        </div>
        <div
          data-testid="legend"
        >
          Legend
        </div>
        <div
          data-testid="line-Temperature"
        >
          Temperature
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MultiParameterChart matches snapshot with two parameters 1`] = `
<div
  class="card p-2 sm:p-6 "
>
  <h3
    class="text-base sm:text-lg font-semibold text-gray-900 dark:text-white mb-2 sm:mb-4 flex items-center"
  >
    <svg
      class="w-5 h-5 mr-2 text-blue-500 dark:text-blue-400"
      fill="none"
      stroke="currentColor"
      viewBox="0 0 24 24"
    >
      <path
        d="M9 19v-6a2 2 0 00-2-2H5a2 2 0 00-2 2v6a2 2 0 002 2h2a2 2 0 002-2zm0 0V9a2 2 0 012-2h2a2 2 0 012 2v10m-6 0a2 2 0 002 2h2a2 2 0 002-2m0 0V5a2 2 0 012-2h2a2 2 0 012 2v14a2 2 0 01-2 2h-2a2 2 0 01-2-2z"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
      />
    </svg>
    Hourly Weather Parameters
  </h3>
  <div
    class="rounded-xl p-1 sm:p-4 relative bg-gradient-to-br from-white via-gray-50 to-white border border-gray-200 shadow-lg backdrop-blur-sm"
    style="min-height: 400px; height: 400px;"
  >
    <div
      data-testid="responsive-container"
    >
      <div
        data-testid="line-chart"
      >
        <div
          data-testid="cartesian-grid"
        >
          CartesianGrid
        </div>
        <div
          data-testid="x-axis"
        >
          XAxis
        </div>
        <div
          data-testid="y-axis"
        >
          YAxis
        </div>
        <div
          data-testid="y-axis"
        >
          YAxis
        </div>
        <div
          data-testid="legend"
        >
          Legend
        </div>
        <div
          data-testid="line-Temperature"
        >
          Temperature
        </div>
        <div
          data-testid="line-Wind Speed (10m)"
        >
          Wind Speed (10m)
        </div>
      </div>
    </div>
  </div>
</div>
`;
